pkg/auth/providers/oidc/oidc_provider.go

```diff
@@ -0,0 +1,418 @@
+	userMGR     user.Manager
+	tokenMGR    *tokens.Manager
+}
+
+func Configure(ctx context.Context, mgmtCtx *config.ScaledContext, userMGR user.Manager, tokenMGR *tokens.Manager) common.AuthProvider {
+	oidcClient := Client{
+		httpClient: &http.Client{
+			Timeout: time.Second * 30,
+		},
+	}
+	return &oidcProvider{
+		ctx:         ctx,
+		authConfigs: mgmtCtx.Management.AuthConfigs(""),
+		secrets:     mgmtCtx.Core.Secrets(""),
+		oidcClient:  &oidcClient,
+		userMGR:     userMGR,
+		tokenMGR:    tokenMGR,
+	}
+}
+
+func (o *oidcProvider) GetName() string {
+	return Name
+}
+
+func (o *oidcProvider) AuthenticateUser(ctx context.Context, input interface{}) (v3.Principal, []v3.Principal, string, error) {
+	login, ok := input.(*v32.OIDCLogin)
+	if !ok {
+		return v3.Principal{}, nil, "", fmt.Errorf("unexpected input type")
+	}
+	return o.LoginUser(login, nil)
+}
+
+func (o *oidcProvider) LoginUser(oauthLoginInfo *v32.OIDCLogin, config *v32.OIDCConfig) (v3.Principal, []v3.Principal, string, error) {
+	var userPrincipal v3.Principal
+	var groupPrincipal []v3.Principal
+	var caPool *x509.CertPool
+	var err error
+
+	if config == nil {
+		config, err = o.getOIDCConfig()
+		if err != nil {
+			return userPrincipal, nil, "", err
+		}
+	}
+	logrus.Debugf("[generic oidc] loginuser: using code to get oauth token")
+	code := oauthLoginInfo.Code
+
+	if config.PrivateKey != "" && config.Certificate != "" {
+		caPool, err = createCACertPool(config.PrivateKey, config.Certificate)
+		if err != nil {
+			return userPrincipal, groupPrincipal, "", err
+		}
+	}
+	logrus.Debugf("[generic oidc] loginuser: get well-known configuration")
+	wk, err := o.oidcClient.getWellKnownConfig(config, caPool)
+	if err != nil {
+		return userPrincipal, groupPrincipal, "", err
+	}
+	logrus.Debugf("[generic oidc] loginuser: validate provider is supported")
+	isSupportedProvider := isSupportedProvider(*wk)
+	if !isSupportedProvider {
+		return userPrincipal, groupPrincipal, "", errors.New("[generic oidc] Provider does not support required scopes and claims")
+	}
+	configUnsetEndpoints(*wk, config)
+	logrus.Debugf("[generic oidc] loginuser: exchanging code for oauth tokens")
+	oauthTokens, err := o.oidcClient.getAccessTokens(code, config, caPool)
+	if err != nil {
+		return userPrincipal, groupPrincipal, "", err
+	}
+	accessToken := oauthTokens.AccessToken
+	userInfo, err := o.oidcClient.getUserInfo(accessToken, config, caPool)
+	if err != nil {
+		return userPrincipal, groupPrincipal, "", err
+	}
+
+	userPrincipal = o.toPrincipal(UserType, userInfo, nil)
+	logrus.Debugf("[generic oidc] loginuser: Checking user's access to Rancher")
+	allowed, err := o.userMGR.CheckAccess(config.AccessMode, config.AllowedPrincipalIDs, userPrincipal.Name, nil)
+	if err != nil {
+		return userPrincipal, groupPrincipal, "", err
+	}
+	if !allowed {
+		return userPrincipal, groupPrincipal, "", httperror.NewAPIError(httperror.Unauthorized, "unauthorized")
+	}
+	return userPrincipal, groupPrincipal, "", nil
```

Comment:
so is there no groupPrincipals found here?

Reply:
So I was able to add the user's groups using the original method we talked about, but there is still no way just to search groups - I am getting groups from the user's info

Reply:
yes groups list from the user info during login only is fine, all SAML providers do that anyway since searching groups is not possible.